Dev/edit rights

somazx opened the pull request:
This pull request replaces the premis_upgrades branch/pull request. 
Includes:
- Schema changes to handle one:many rightsBasis:rightsGranted and is premis 2.x friendly.
- new forms for editing creating rightsBasis with n rightsGranted attached

somazx committed:
Creating form for rights management.

somazx committed:
added js for form template

somazx committed:
Handling of new rights now works.

"Validates" and saves it associated to the resource.

somazx committed:
turn on autocomplete support

somazx committed:
temporary link to edit existing rights

somazx committed:
Form should just submit to the existing reqst url

somazx committed:
handled editing rights

somazx committed:
Updated Schema for one:many basis:act

somazx committed:
handling embedded forms for each 'act'

somazx committed:
store a 'blank' form instead of a 1000'th subform

somazx committed:
working javascript to add items

somazx committed:
rights object saves relatedRights

somazx committed:
updated to show related grantedRights

somazx committed:
build title based on act and restriction values

somazx committed:
handle saving grantedRights submitted via the form

somazx committed:
better routing, link to create to right

somazx committed:
fixed counter

somazx committed:
fixed link

somazx committed:
improved finding pre-existing grantedRights

somazx committed:
added granted right notes field

somazx committed:
date widgets now use date for input type

somazx committed:
formatting fixes

somazx committed:
removed attempt to render act/restriction

this is just so the template will at least not cause a hard error

somazx committed:
added missing notes

somazx committed:
fixed dates not being null'd with empty date value

somazx committed:
changed wording

somazx committed:
removing restriction, act_id

somazx committed:
removed use of reference

somazx committed:
just grab all columns in rights for now

somazx committed:
added divider before create new rights

somazx committed:
added help texts

somazx committed:
activated description javascript and edit css

somazx committed:
tweak/hack to how description.js works

if it looks like description-bottom is to be used, first check if there
is room to the left of #content and assume we can render there.

somazx committed:
tweak to offsetLeft value

somazx committed:
added more help

somazx committed:
fixed help for embedded form (granted rights)

embedded form fields are instances of sfFormFieldSchema no sfForm, and
thus don't have a help() method. fortunately you can pass in the help
text to renderRow without overriding previously set value (passing in
null)

somazx committed:
never used code

somazx committed:
making method protected

somazx committed:
renaming classes to be singular 'right'

to be consistent with routes

somazx committed:
adding action for deleting a Right

somazx committed:
script to handle delete A right button/trigger

somazx committed:
enabling delete button

somazx committed:
delete related grantedRights

somazx committed:
only allow edit/delete of Rights on the immediate object

somazx committed:
Indicate inherited vs related right

somazx committed:
removing junk; accident commit

somazx committed:
handle deleting of granted rights

djjuhasz commented on apps/qubit/modules/informationobject/actions/deleteRightAction.class.php:
```diff
@@ -0,0 +1,31 @@
+  public function execute($request)
+  {
+    $this->setLayout(null);
+    $this->setTemplate(null);
```
I don't think this is necessary;  AFAIK, using return $this->renderText() is sufficient to prevent trying to load a template page.

somazx replied:
Fixed. Thanks!

djjuhasz commented on apps/qubit/modules/informationobject/actions/deleteRightAction.class.php:
```diff
@@ -0,0 +1,31 @@
+
+    $this->right = $this->getRoute()->resource;
+    $this->right->delete();
+    return $this->renderText(json_encode(true));
```
Should be a blank line above return statement

somazx replied:
Fixed. Thanks!

djjuhasz commented on apps/qubit/modules/informationobject/actions/deleteRightAction.class.php:
```diff
@@ -0,0 +1,31 @@
+{
+  public function execute($request)
+  {
+    $this->setLayout(null);
```
I think $this->getResponse()->setContentType('application/json'); is preferable to prevent loading the page decorations.

sevein replied:
``` php
return sfView::NONE;
```

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,370 @@
+  {
+    $widget = new sfWidgetFormInput;
+    $widget->setAttribute('type', 'date');
+    return $widget;
```
Blank line before return

somazx committed:
Added blank line before return.

somazx committed:
Removed set Layout/Template; not necessary.
They're not needed when using return renderText

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,370 @@
+        $this->form->setValidator('endDate', new sfValidatorString);
+        $this->form->setWidget('endDate', $this->dateWidget());
+        $this->form->getWidgetSchema()->endDate->setLabel($this->context->i18n->__('End'));
+        break;
```
Blank line before break;

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,370 @@
+            $this->form->setDefault('basis', $this->context->routing->generate(null, array($item, 'module' => 'term')));
+          }
+          $choices[$this->context->routing->generate(null, array($item, 'module' => 'term'))] = $item->__toString();
+        }
```
Blank line after closing brace "}"

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,370 @@
+          if($data['delete'] === "true")
+          {
+            $grantedRight->delete();
+            continue;
```
blank line before continue;

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,370 @@
+
+          $grantedRight->act            = $actparams['_sf_route']->resource;
+          $grantedRight->restriction    = $data['restriction'];
+          // empty dates come in as empty strings, but propel wants 'null' or it'll default to today's date
```
Add blank line before comment

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,370 @@
+            $this->right->grantedRights[] = $grantedRight;
+          }
+        }
+      case 'blank':
```
Add blank line before case

somazx committed:
Code style edits/fixes

somazx committed:
Code style edits/fixes

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,379 @@
+          }
+          
+          // none found, so make a new one
+          if(! $grantedRight)
```
Explicitly test "not" condition e.g. if (null === $grantedRight).  Also note space between "if" and opening parenthesis

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,379 @@
+          // if one was found, but user 
+          // has requested it be deleted
+          // then lets delete it.
+          if($data['delete'] === "true")
```
Add space between "if" and opening parenthesis "if (".  Constant value should come before variable value in comparison statement, i.e. 

if ("true" === $data['delete'])

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,379 @@
+          $grantedRight->notes          = $data['notes'];
+
+          // relate it to the Right if it is new
+          if(! $grantedRight->id)
```
if (! isset($grantedRight->id)) ?

djjuhasz replied:
Nope, I'm wrong. Should be:

if (null === $grantedRight->id)

ref: https://www.qubit-toolkit.org/wiki/Coding_standard#is_null.28.29

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,379 @@
+      $this->relation->object = $this->right;
+      $this->relation->typeId = QubitTerm::RIGHT_ID;
+      $this->relation->subject = $this->resource;
+      $this->relation->save();
```
Blank line before ->save();

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,379 @@
+  {
+    $results = QubitRelation::getRelatedSubjectsByObjectId('QubitInformationObject', $id);
+
+    return $results[0];
```
I think this will throw a warning if $results == null ?

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,379 @@
+        $this->right = $resource;
+
+        break;
+      case 'QubitInformationObject':
```
blank line before case

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,379 @@
+    }
+  }
+
+  protected function getObject($id)
```
This is grabbing the QubitInformationObject related to the Right?  I would prefer the method name to be more explicit, e.g. getRelatedInformationObject because we have a QubitObject class and `object` SQL table.

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,379 @@
+    $resource = $this->getRoute()->resource;
+    $type = get_class($resource);
+
+    switch ($type) {
```
Because $type isn't reused, more succinct to use switch(get_class($resource)) and saves an extra varable

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,379 @@
+    $subForm->getValidatorSchema()->setOption('allow_extra_fields', true);
+    if( sizeof($this->right->grantedRights)  )
+    {
+      foreach ($this->right->grantedRights as $i => $gr) {
```
Newline before opening brace

djjuhasz replied:
We prefer $item for assigning foreach values, e.g.
foreach ($this->right->grantedRights as $i => $item)

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,379 @@
+      foreach ($this->right->grantedRights as $i => $gr) {
+        $subForm->embedForm($i, $this->grantedRightFormSetup($gr));
+      }
+    } else {
```
should be:
}
else
{

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,379 @@
+      }
+    } else {
+      $gr = new QubitGrantedRight;
+      $subForm->embedForm(0, $this->grantedRightFormSetup($gr));
```
Does $subForm->embedForm(0, $this->grantedRightFormSetup(new QubitGrantedRight));  work?

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,379 @@
+    // and generate an act row for each one.
+    $subForm = new sfForm();
+    $subForm->getValidatorSchema()->setOption('allow_extra_fields', true);
+    if( sizeof($this->right->grantedRights)  )
```
should be:
if (0 < count($this->right->grantedRights))

See:
https://www.qubit-toolkit.org/wiki/Coding_standard#count.28.29_vs._sizeof.28.29
https://www.qubit-toolkit.org/wiki/Coding_standard#empty.28.29

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,379 @@
+  protected function grantedRightFormSetup($grantedRight)
+  {
+    // if new, unsaved right, then id can be set to 0
+    $grantedRight->id = (integer) $grantedRight->id;
```
Would prefer to remove this line and test if (null === $grantedRight->id) to determine if it's a new granted right.

somazx committed:
Code style edits/fixes

djjuhasz commented on apps/qubit/modules/informationobject/actions/editRightAction.class.php:
```diff
@@ -0,0 +1,379 @@
+
+  protected function getObject($id)
+  {
+    $results = QubitRelation::getRelatedSubjectsByObjectId('QubitInformationObject', $id);
```
Qualify the relationship type in case there are other objects related to this QubitRight:
$results = QubitRelation::getRelatedSubjectsByObjectId('QubitInformationObject', $id, array('typeId' => QubitTerm::RIGHT_ID));

somazx committed:
Fixing delete button.
fixed: a href wasn't specified
fixed: return false so href isn't followed
fixed: javascript not being loaded

somazx committed:
Code styling fixes. Function renamed to be less vague.

somazx committed:
Code styling fixes. Function renamed to be less vague.

somazx committed:
Removing anything to do with rights management

somazx committed:
Refactor edit/deleteRight action into rights module

Rights are independent of the informationobject and exist on other objects such as Accession objects and and Digital Objects.

somazx committed:
Rrights CRUD now works for digitalobjects and information objects

somazx committed:
Rights CRUD now works for accession objects

somazx committed:
Added identifierType, Value & Role to Rights
Also moved custom save method out of BaseRights into QubitRights to avoid it being overwritten anytime someone runs symfony propel:model-build

somazx committed:
Adding Migration for new rights.

somazx committed:
Migration handles identifier type,value, role.

somazx committed:
added migration to MigrateTask class

somazx committed:
fix to dropping act_id foreign key
